libs/gltfio/src/DependencyGraph.h

```diff
@@ -0,0 +1,108 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GLTFIO_DEPENDENCY_GRAPH_H
+#define GLTFIO_DEPENDENCY_GRAPH_H
+
+#include <utils/Entity.h>
+
+#include <tsl/robin_map.h>
+#include <tsl/robin_set.h>
+
+#include <queue>
+#include <string>
+
+namespace filament {
+    class MaterialInstance;
+    class Texture;
+}
+
+namespace gltfio {
+
+/**
+ * Internal graph that enables FilamentAsset to discover "ready-to-render" entities by tracking
+ * the Texture objects that each entity depends on.
+ *
+ * Renderables connect to a set of material instances, which in turn connect to a set of parameter
+ * names, which in turn connect to a set of texture objects. These relationships are not easily
+ * inspectable using the Filament API or ECS.
+ *
+ * One graph corresponds to a single glTF asset. The graph only contains weak references, it does
+ * not have ownership over any Filament objects. Here's an example:
+ *
+ *    Entity          Entity   Entity   Entity
+ *      |            /      \     |     /
+ *      |           /        \    |    /
+ *   Material   Material       Material
+ *             /   |    \          |
+ *            /    |     \         |
+ *        Param  Param   Param   Param
+ *           \     /        \     /
+ *            \   /          \   /
+ *          Texture         Texture
+ *
+ * Note that the left-most entity in the above graph has no textures, so it becomes ready as soon as
+ * finalize is called.
+ */
+class DependencyGraph {
+public:
+    using Material = filament::MaterialInstance;
+    using Entity = utils::Entity;
+
+    // Returns a ready-to-render entity, or 0 if no new entities have become renderable.
+    Entity popReadyRenderable() noexcept;
+
+    // These are called during the initial asset loader phase.
+    void addEdge(Entity entity, Material* material);
+    void addEdge(Material* material, const char* parameter);
+
+    // This is called at the end of the initial asset loading phase.
+    void finalize();
+
+    // These are called after textures have created and decoded.
+    void addEdge(filament::Texture* texture, Material* material, const char* parameter);
+    void markAsReady(filament::Texture* texture);
+
+private:
+    struct TextureStatus {
+        filament::Texture* texture;
+        bool ready;
+    };
+
+    struct MaterialStatus {
+        tsl::robin_map<std::string, TextureStatus*> params;
+    };
+
+    struct EntityStatus {
+        tsl::robin_set<Material*> materials;
+        size_t numReadyMaterials = 0;
+    };
+
+    void markAsReady(Material* material);
+    TextureStatus* getStatus(filament::Texture* texture);
+
+    tsl::robin_map<Entity, EntityStatus> mEntityToMaterial;
+    tsl::robin_map<Material*, tsl::robin_set<Entity>> mMaterialToEntity;
+    tsl::robin_map<Material*, MaterialStatus> mMaterialToTexture;
```

Comment:
Maybe rename to `mMaterialToStatus`? Or simply `mMaterials` to be consistent with `mTextures`.

Reply:
This is named this way because it holds the directed edges in the DAG that (indirectly) go from materials to textures.  The mTextures field is different; it's just a bag of TextureStatus objects that are shared from various parts of the graph.

I agree that EntityStatus, MaterialStatus, and TextureStatus are awkward names.  Maybe I should rename them to EntityNode, MaterialNode, and TextureNode to better emphasize their role in the DAG.